src/PostgREST/DbStructure.hs

```diff
@@ -451,6 +499,11 @@ allTables =
       n.nspname AS table_schema,
       c.relname AS table_name,
       NULL AS table_description,
+      (
+        has_table_privilege(c.oid, 'INSERT,UPDATE,DELETE,TRUNCATE')
+        or has_any_column_privilege(c.oid, 'INSERT, UPDATE')
+      )
```

Comment:
I don't think this bit should be added here. That's in the `allTables` query, which should not be concerned about permissions - just about the general insertability, updatability and deletability. Those values should then be used in the `OPTIONS` request in the `Allow` header: https://tools.ietf.org/html/rfc7231#section-7.4.1

I read this as the returned values should only depend on the methods supported by the entity - not limited to those with privileges for the current user. @steve-chavez WDYT?

Reply:
I don't even think it will work here as expected, because those privileges will only be checked for the authenticator role.

Reply:
> I read this as the returned values should only depend on the methods supported by the entity - not limited to those with privileges for the current user.

@wolfgangwalther But according to your comment here https://github.com/PostgREST/postgrest/issues/1647#issuecomment-731846689, we do need to look at the privileges for restricting certain http methods, right?

Hm, besides restricting OPTIONS, would we also need to disallow HTTP methods and respond with `405 Method Not Allowed`? That would avoid hitting the db in case of missing privileges, but I'm not sure if we should do it. It would make development a bit more complicated - the user would have to refresh PostgREST schema cache when changing privileges on the db.

Perhaps the `postgrest-contrib` approach mentioned on https://github.com/PostgREST/postgrest/pull/1710#issuecomment-808788905 would be more fit for users that want to restrict the http methods.

Reply:
> That's in the allTables query, which should not be concerned about permissions

Yes, I had doubts about modifying it. Won't add privileges check here.

Reply:
> @wolfgangwalther But according to your comment here [#1647 (comment)](https://github.com/PostgREST/postgrest/issues/1647#issuecomment-731846689), we do need to look at the privileges for restricting certain http methods, right?

No. The whole comment is about `OPTIONS`, which is handled by `handlerInfo`. This does not access the database at all, so it's role-independent. The question here is not whether insert/update/delete are **allowed**, but if they are technically **possible** at all. This is only relevant for views, that are not auto-updateable. In this case, the possibilty to do anything is dependent on the availability of a `INSTEAD OF` trigger.

> Hm, besides restricting OPTIONS, would we also need to disallow HTTP methods and respond with 405 Method Not Allowed? That would avoid hitting the db in case of missing privileges, but I'm not sure if we should do it. It would make development a bit more complicated - the user would have to refresh PostgREST schema cache when changing privileges on the db.

Hm. I think any privilege-related thing needs to be determined at run-time, not by schema cache. We're not keeping track of different schema caches for different users. Even the OpenAPI output is making a separate query on each request to determine the current permissions. So if you want to take permissions into account, you'd need to hit the database anyway. But looking at just the technical possibility to do an `UPDATE`, `DELETE` or `INSERT` - yeah, we could block those before they even get to the database - and that would not be as inconveniet, because you'd need to refresh the schema cache after adding a trigger. That seems reasonable. I suggested a refactor of our whole routing somewhere else - this would be easily included in that kind of refactor. So I would not go for it independently right now - not worth the effort right now.

Reply:
> > That's in the allTables query, which should not be concerned about permissions
> 
> Yes, I had doubts about modifying it. Won't add privileges check here.

Following my argument above, we should not add them anywhere. Otherwise the fields on `Table` would have different meaning depending on whether the query was made with `accessibleTables` or `allTables` - that seems really odd.

Reply:
> Ah, got it. So if it's about INSTEAD OF triggers, then only querying `information_schema.triggers`(or an internal pg table) would be needed right? No need to look into the view's privileges.

Yeah, the query for that is all sorted (it worked fine for `INSERT` before and just needed to be adapted for `UPDATE` or `DELETE`). My question was actually not about the PG side of things but only about the best HTTP semantics here - basically whether my interpretation of the RFC I cited above was correct or not:

Should an `OPTIONS` request return the `Allow` header dependent on user privileges or not? My answer was no - do you agree?

Reply:
> Should an OPTIONS request return the Allow header dependent on user privileges or not? My answer was no - do you agree?

Agree. This reminds me of this [comment from Roy Fielding](https://roy.gbiv.com/untangled/2008/rest-apis-must-be-hypertext-driven#comment-732):

> HTTP operations are generic: they are allowed or not, per resource, but they are always **valid**. Hypertext doesn’t usually tell you all the operations allowed on any given resource; it tells you which operation to use for each potential transition.

And the [Allow RFC](https://tools.ietf.org/html/rfc7231#section-7.4.1) says:

> The purpose of this field is strictly to inform the recipient of **valid** request methods associated with the resource.

So yes, we're informing the recipient of the **valid** methods. Allowance is defined at run-time and if the resource replies with 401, then the recipient could request higher privileges.

This also makes me think about RPC, it should not include GET(when it's not IMMUTABLE/STABLE) to avoid `on a read-only transaction` errors. Actually I've realized that right now RPC doesn't allow OPTIONS.

Reply:
> This also makes me think about RPC, it should not include GET(when it's not IMMUTABLE/STABLE) to avoid `on a read-only transaction` errors. Actually I've realized that right now RPC doesn't allow OPTIONS.

Ah good point. And it's not too easy to add, either, because of overloaded functions. However, with the routing-refactor it should be easily doable, too. Just need to make sure we keep track of that!